nodejs/week2/app.js

```diff
@@ -0,0 +1,105 @@
+import express from "express";
+import { readFile } from "fs/promises";
+
+const app = express();
+const port = process.env.PORT || 3000;
+
+app.use(express.json());
+
+app.get("/", (req, res) => {
+  res.send("This is a search engine");
+});
+
+// GET /search
+app.get("/search", async (req, res) => {
+  const { q } = req.query;
+
+  try {
+    const filePath = new URL("./documents.json", import.meta.url);
+    const contents = await readFile(filePath, { encoding: "utf8" });
+    const documents = JSON.parse(contents);
+
+    if (!q) {
+      return res.json(documents);
+    } else {
+      const filteredDocs = documents.filter((doc) =>
+        Object.values(doc).some(
+          (value) =>
+            value !== undefined &&
+            value.toString().toLowerCase().includes(q.toLowerCase())
+        )
+      );
+      return res.json(filteredDocs);
+    }
+  } catch (err) {
+    console.error(err.message);
+  }
+});
+
+// GET /documents/:id
+app.get("/documents/:id", async (req, res) => {
+  const id = parseInt(req.params.id);
+  if (isNaN(id)) {
+    return res.status(400).send("Invalid Id");
+  }
+  try {
+    const filePath = new URL("./documents.json", import.meta.url);
+    const contents = await readFile(filePath, { encoding: "utf8" });
+    const documents = JSON.parse(contents);
+  
+    const filteredDoc = documents.find((doc) => doc.id === id);
+
+    if (filteredDoc) {
+      return res.json(filteredDoc);
+    } else {
+      return res.status(404).send("Document Not Found");
+    }
+  } catch (err) {
+    console.error(err.message);
+  }
+});
+
+// POST /search
+app.post("/search", async (req, res) => {
+  const { q } = req.query;
+  const { fields } = req.body;
+
+  if (q && fields) {
+    return res
+      .status(400)
+      .send("Cannot use both query parameter and request body.");
+  }
```

Comment:
I see, so there are two options available. In that case I would keep the conditions logic in particular endpoints and extract the actual functions (below) to separate utility functions. You know, to keep stuff DRY (Don't Repeat Yourself) :)

Reply:
I've already extract the functions into separate utilities. You're absolutely right, it's good clean code practice :D. Thank u so much for your review!🫶